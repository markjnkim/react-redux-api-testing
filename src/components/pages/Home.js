import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import get from 'lodash.get';
import { locate } from '../../actions';
import data from '../../fetch/locObj'
import logo from '../../assets/logo.png';
import ListingCards from '../listing_cards';
import Locate from '../../icons/locate';
import MapPin from '../../icons/map-pin';
import Delivery from '../../icons/delivery';
import Dispensary from '../../icons/dispensary';
import Doctor from '../../icons/doctor';
import {
  AppHeader,
  AppWrapper,
  AppContent,
  ListingGroups,
  HeroSection,
  ContentContainer,
  LocationSection,
  TextContent,
  LocateButton,
  CardStyle,
} from '../styles';

const regionTypes = ['delivery', 'dispensary', 'doctor'];
const regionLabels = {
  delivery: <span><Delivery fill={'#7e7979'} width={'40px'} height={'25px'} />  Delivery Services</span>,
  dispensary: <span><Dispensary fill={'#7e7979'} width={'40px'} height={'25px'} />  Dispensary Storefronts </span>,
  doctor: <span><Doctor fill={'#7e7979'} width={'40px'} height={'25px'} />  Doctors</span>,
};

export class App extends Component {
  constructor(props) {
    super(props);
    console.log("Constructor: ", this.props)
    this.state = {
      loadingTimer: 0,
    };
  }
  componentDidMount() {
    // this.props.requestLectures(this.props.isLoading);

    console.log("CDM: ", this);
  }
  locateMe() {
  const { dispatch } = this.props;
  if (navigator.geolocation) {
    navigator.geolocation.getCurrentPosition(position => {
        dispatch(locate(position.coords));
      });
    }
  };

  render() {
    const { isLocating, location, regions, error } = data.data;
    console.log("Props Location: ", location)
    
    const getLabel = (listings, label) => {
      if (get(listings, 'listings').length) {
        return (
          <div key={label}>
            <strong> {label} </strong>
          </div>
        );
      }
      return <div />;
    };

    return (
      <AppWrapper>
        <AppHeader>
          <img src={logo} alt="weedmaps logo" />
        </AppHeader>
        <HeroSection>
          <ContentContainer>
            <LocationSection>
              <h2>
                <MapPin fill={'#7e7979'} width={'60px'} height={'40px'} />
                <span> {location ? location.name : ''} </span>
                <span> {isLocating && !location ? '...locating' : ''} </span>
              </h2>
              <LocateButton onClick={this.locateMe.bind(this)}>
                <Locate fill={'#7e7979'} />
                <span> Locate Me </span>
              </LocateButton>
            </LocationSection>
            <TextContent>
              Lorem Ipsum dolor sit amet, consectetur adispiscing elit, sed do
              eiusmod tempor incididunt ut labore et dolore magna aligqua. Ut
              enim ad minim veniam, quis.
            </TextContent>
          </ContentContainer>
        </HeroSection>
        <AppContent>
          {error && <div> {error.message} </div>}
          {regions && (
            <React.Fragment>
              {regionTypes.map(regionType => (
                <ListingGroups key={regionType}>
                  <h2>
                    {getLabel(regions[regionType], regionLabels[regionType])}
                  </h2>
                  <CardStyle>
                    <ListingCards
                      listings={get(regions[regionType], 'listings')}
                    />
                  </CardStyle>
                </ListingGroups>
              ))}
            </React.Fragment>
          )}
          {/* } */}
        </AppContent>
      </AppWrapper>
    );
  }
}

const mapStateToProps = (state) => {
  console.log("MSTP :", state);
  return { location: state.location };
}


App.propTypes = {
  isLocating: PropTypes.bool.isRequired,
  location: PropTypes.object,
  regions: PropTypes.object,
  dispatch: PropTypes.any,
  error: PropTypes.object,
};

App.defaultProps = {
  isLocating: false,
  location: {},
  regions: {},
  error: {},
};

export default connect(mapStateToProps)(App);
