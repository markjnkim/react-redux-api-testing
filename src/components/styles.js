import styled from "styled-components";


export const AppHeader = styled.div`
  height: 70px;
  display: flex;
  padding: 0 20px;
  justify-content: space-between;
  align-items: center;
  background-color: #222;
  color: #fff;

  img {
    width: 110px;
    height: 25px;
  }
`;

export const AppWrapper = styled.div`
  margin-bottom: 20px;
  text-align: center;
  background-color: #EAEAEA;
  min-width: 450px !important;
`;

export const Icon = styled.div`
  // padding-top: 40px;
  // margin-top: 10px;
  display: inline-block;
  // align: bottom;
  position: absolute:
  bottom: 0px;
`;

export const AppContent = styled.div`
  width: 90%;
  max-width: 1200px;
  margin: 10px auto;
`;

export const ListingGroups = styled.div`
  margin-top: 30px;
  h2 {
    text-align: left;
  }
`;

export const HeroSection = styled.div`
  background-color: white;
  display: flex;
  justify-content: center;
  width: 100%;
  height: auto;
  border-bottom: 1px solid #e5e4e4;
  padding: 0 0 16px;

  h2 {
    font-weight: 300;
    color: #a4a2a2;
    display: flex;
    align-items: center;
    span {
      margin-left: 5px;
    }
  }
`;

export const ContentContainer = styled.div`
  width: 100%;
  max-width: 1200px;
  padding: 16px;
`;

export const LocationSection = styled.div`
  display: flex;
  align-items: center;
  width: 100%;
  justify-content: space-between;
  height: 80px;
`;

export const TextContent = styled.div`
  width: 100%;
  max-width: 1100px;
  height: auto;
  text-align: left;
  color: #a4a2a2;
  line-height: 25px;
  font-size: 14px;
  letter-spacing: 1px;
  font-weight: 300;
`;

export const LocateButton = styled.a`
  width: 105px;
  height: 25px;
  display: flex;
  align-items: center;
  justify-content: start;
  border: 1px solid #c2baba;
  color: #7e7979;
  border-radius: 3px;
  padding: 5px;
  font-size: 12px;
  cursor: pointer;
  text-transform: uppercase;

  svg {
    margin-right: 5px;
  }
`;

export const CardStyle = styled.div`
  // margin: 4px;
  // padding: 3px;

  @media (max-width: 600px) {
    text-align: center;
  }
`;

export const DIV1 = styled.div`
  text-align: left;
  margin-left: 7px;
  font-weight: bold;

  @media (max-width: 600px) {
    font-size: 12px !important;
  }
`;

export const DIV2 = styled.div`
  margin: 4px;
  padding: 3px;
  color: grey;
  font-size: 14px;

  @media (max-width: 600px) {
    font-size: 10px !important;
  }
`;

export const DIV3 = styled.div`
  margin: 4px;
  padding: 3px;
  color: grey;
`;

export const DivWrapper = styled.div`
  margin: 4px;
  padding: 3px;
  order: 2;
  @media (max-width: 700px) {
    order: 1;
    float: left;
  }
`;

export const CardWrapper = styled.div`
  width: 80%;
  height: 70%;
  display: flex;
  text-align: left;
  margin-top: 20px;
  margin-bottom: 20px;
  margin-right: 20px;
  padding: 0;
  background-color: white;
  border-radius: 6px;
  cursor: pointer;

  @media (max-width: 700px) {
    width: 100%;
  }
  // @media (max-width: 600px) {
  //   width: 92%;
  // }
  // @media (max-width: 500px) {
  //   width: 99%;
  // }
`;

export const AvatarWrapper = styled.div`
  vertical-align: bottom;
  padding: 1%;
  order: 1;
  @media (max-width: 500px) {
    padding-bottom: 19px;
  }
`;

export const SpanRating = styled.span`
  color: grey;
  padding-left: 5px;
  font-size: 13px;
  
  @media (max-width: 700px) {
    order: 2;
  }
`;
export const SPAN = styled.span`
  color: #12C3A0;
  @media (max-width: 700px) {
    order: 3;
    display: none;
  }
`;

export const DIVO = styled.div`
  color: #12C3A0;
  display: none;
  @media (max-width: 700px) {
    order: 4;
    position: relative;
    display: block;
    padding-left: 10px;
    font-size: 11px;
  }
`;

export const DetailHeader = styled.div`
  height: 230px;
  display: flex;
  padding: 0 20px;
  justify-content: space-between;
  align-items: center;
  background-color: #222;
  color: #fff;
  clearfix: auto;

  img {
    width: 210px;
    height: 225px;
  }
`;

export const DIVNAME = styled.div`
text-align: left;
margin-left: 7px;
font-weight: bold;
font-size: 3vw;

@media (max-width: 600px) {
  font-size: 12px !important;
}
`;

export const SpanRatingDetail = styled.span`
  color: #12C3A0;
  padding-left: 5px;
  font-size: 3vw;
  
  @media (max-width: 700px) {
    order: 2;
  }
`;

export const AppWrapperDetail = styled.div`
  margin-bottom: 20px;
  text-align: center;
  background-color: #EAEAEA;
  min-width: 450px !important;
  height: 100%;
  min-height: 600px;
`;