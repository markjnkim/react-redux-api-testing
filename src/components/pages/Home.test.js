import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import sinon from 'sinon';

import * as actions from '../../actions';
import { Home } from './Home';
import { LocateButton } from '../styles';
import * as coords from '../../__test__/mocks/coord-mock.json';
import * as location from '../../__test__/mocks/location-mock.json';

// let [requestLocation, receiveLocation] = new Array(3).fill(jest.fn());

function shallowSetup() {
  // Sample props to pass to our shallow render
  const props = {
    listing: {
      selectedListing: null
    },
    location: {
      isLocating: false,
      location: location,
      regions: null,
    },
    dispatch: jest.fn(),
  }
  // wrapper instance around rendered output
  const enzymeWrapper = shallow(<Home {...props} />);

  return {
    props,
    enzymeWrapper
  };
}

describe('Shallow rendered Home', () => {
  const { enzymeWrapper, props } = shallowSetup();
  it('renders without crashing given the required props', () => {
    expect(toJson(enzymeWrapper)).toMatchSnapshot()
  });

  it('should render a Home with state', () => {
    // Setup wrapper and assign props.
    const { enzymeWrapper, props } = shallowSetup();
    // enzymeWrapper.find(selector) : Find every node in the render tree that matches the provided selector. 
    expect(enzymeWrapper.find('span').first().text()).toBe("  ");
    expect(enzymeWrapper.containsMatchingElement(<span>props.location.name</span>)).toBe(false);
    expect(enzymeWrapper.containsMatchingElement(<span>location.name</span>)).toBe(false);
    expect(enzymeWrapper.containsMatchingElement(<span> Locate Me </span>)).toBe(true);
  });

  it('sets the locate prop as the `value` prop on the LocateButton component', () => {
    const props = {
      listing: {
        selectedListing: null
      },
      location: {
        isLocation: false,
        location: null,
        regions: null,
      },
      dispatch: jest.fn(),
    }
  //   const wrapper = shallow(<App {...props} />)
  //   // Query for the Picker component in the rendered output
  //   const PickerComponent = wrapper.find(Picker)
  //   expect(PickerComponent.props().value).toBe(props.selectedSubreddit)
  });

  it('locate button dispatches the correct actions', () => {
    const props = {
      listing: {
        selectedListing: null
      },
      location: {
        isLocation: false,
        location: null,
        regions: null,
      },
      dispatch: jest.fn(),
    }
    // Mock event to be passed to the locateMe function
    const mockEvent =  {
      preventDefault: jest.fn()
    }
    // Mock the actions we expect to be called
    // let [mockEvent, actions.requestLocation, actions.receiveLocation,actions.error ] = new Array(3).fill(jest.fn());
    actions.requestLocation = jest.fn();
    actions.receiveLocation = jest.fn();
    actions.error = jest.fn();
    const spy = jest.fn();
    
    const wrapper = shallow(<Home {...props} />)
    // Call the function on the component instance, passing the mock event
  //  expect(wrapper.find(LocateButton).length).toBe(undefined);
    // expect(mockEvent.preventDefault).toHaveBeenCalled();
    expect(props.dispatch.mock.calls.length).toBe(0);
    expect(actions.requestLocation.mock.calls.length).toBe(0);
    expect(actions.receiveLocation.mock.calls.length).toBe(0);

  });

  it('locate button clicks', () => {
    // const props = {
    //   listing: {
    //     selectedListing: null
    //   },
    //   location: {
    //     isLocation: false,
    //     location: null,
    //     regions: null,
    //   },
    //   dispatch: jest.fn(),
    // }
    // Mock event to be passed to the locateMe function
    const mockEvent =  {
      preventDefault: jest.fn()
    }
    // Mock the actions we expect to be called
    // let [mockEvent, actions.requestLocation, actions.receiveLocation,actions.error ] = new Array(3).fill(jest.fn());
    // actions.requestLocation = jest.fn();
    // actions.receiveLocation = jest.fn();
    // actions.error = jest.fn();
    // const spy = jest.fn();
    
    const wrapper = shallow(<Home {...props} />)
    // Call the function on the component instance, passing the mock event
  //  expect(wrapper.find(LocateButton).length).toBe(undefined);
    // expect(mockEvent.preventDefault).toHaveBeenCalled();
    const button = wrapper.find(LocateButton);
    button.simulate('click');
    // expect(wrapper.dispatch.mock.calls.length).toBe(0);
    // expect(wrapper.instance().locateMe.toHaveBeenCaled())).;
    expect(actions.receiveLocation.mock.calls.length).toBe(0);

  });
})

describe('Home Locate Button', () => {
  let wrapper, props_;
  beforeEach(() => {
    // spy on the component handleOpen method
    sinon.spy(Home.prototype, "locateMe");
    const { enzymeWrapper, props } = shallowSetup();
    wrapper = enzymeWrapper;
    props_ = props;
  });
  afterEach(() => {
    Home.prototype.locateMe.restore();
  });
})



