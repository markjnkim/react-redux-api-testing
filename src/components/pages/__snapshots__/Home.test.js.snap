// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Location coordinates test renders without crashing given the required props 1`] = `
<styled.div>
  <styled.div>
    <img
      alt="weedmaps logo"
      src="logo.png"
    />
  </styled.div>
  <styled.div>
    <styled.div>
      <styled.div>
        <h2>
          <MapPin
            fill="#7e7979"
            height="40px"
            width="60px"
          />
          <span>
             
             
          </span>
          <span>
             
             
          </span>
        </h2>
        <styled.a
          onClick={[Function]}
        >
          <Locate
            fill="#7e7979"
            height="20px"
            width="20px"
          />
          <span>
             Locate Me 
          </span>
        </styled.a>
      </styled.div>
      <styled.div>
        Lorem Ipsum dolor sit amet, consectetur adispiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aligqua. Ut enim ad minim veniam, quis.
      </styled.div>
    </styled.div>
  </styled.div>
  <styled.div />
  <styled.div />
  <styled.div>
    <img
      alt="weedmaps logo"
      src="logo.png"
    />
  </styled.div>
</styled.div>
`;

exports[`Shallow rendered Home renders without crashing given the required props 1`] = `
<styled.div>
  <styled.div>
    <img
      alt="weedmaps logo"
      src="logo.png"
    />
  </styled.div>
  <styled.div>
    <styled.div>
      <styled.div>
        <h2>
          <MapPin
            fill="#7e7979"
            height="40px"
            width="60px"
          />
          <span>
             
             
          </span>
          <span>
             
             
          </span>
        </h2>
        <styled.a
          onClick={[Function]}
        >
          <Locate
            fill="#7e7979"
            height="20px"
            width="20px"
          />
          <span>
             Locate Me 
          </span>
        </styled.a>
      </styled.div>
      <styled.div>
        Lorem Ipsum dolor sit amet, consectetur adispiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aligqua. Ut enim ad minim veniam, quis.
      </styled.div>
    </styled.div>
  </styled.div>
  <styled.div />
  <styled.div />
  <styled.div>
    <img
      alt="weedmaps logo"
      src="logo.png"
    />
  </styled.div>
</styled.div>
`;
